Exceptions rework

batiati opened the pull request:
Java has 3 semantics for exceptions:

- Checked exceptions, that must be declared on the "throws" block and are required to be explicitly handled by the user, who must decide how to recover from.

- RuntimeExceptions, that can occur somehow due to unexpected runtime values and may be caught on the site or bubbled up to the caller

- Errors, that have similar semantics to "panic", and denote that there is no possible way to recover from.

This PR removes a lot of exceptions wrongly declared on the "throws" block and assures that we are following the correct semantics in exceptions we throw in our public API.

batiati committed:
Exceptions rework

- Better ergonomics when catching exceptions, making all checked exceptions we throw inherit from Exception class.

- Removed unnecessary "throws" block for unchecked exceptions

- New "AssertionError" is thrown on any unexpected condition detected during the execution.

This approach of throwing an "Error" instead of an "Exception", makes explicit to the user that this result is unexcepted and should only be logged and handled at a high level.

batiati committed:
More exception rework

- InitializationException is now an unchecked exception

Rationale: It reduces a lot of friction by not requiring the user to explicitly handle an exception that cannot be recovered from.

Also, the INVALID_ADDRESS result is now handled as IllegalArgumentException, along with other argument validations, making it more idiomatic.

- Removed InterruptedException from the "throws" block.

Rationale: We don't provide ways for the user to cancel an ongoing request, so there is no point for them to explicitly handle this exception.

Actually, this exception is not expected to be thrown ever, the only way for it to happen is somehow by Thread.interrupt() being called. In this case, we abort throwing an AssertionError

- Minor code cleanup, removing unused imports.

batiati committed:
More exceptions rework

- Preserves the Future<T> interface by not removing the InterruptedException from there.
Rationale: This is a valid use case of InterruptedException that was not covered by the previous commit.
Some async framework could use this exception to interrupt and resume a Future<T> anytime.
Our sync API now is using the "waitForCompletionUninterruptibly" and still assumes that no interruption will ever occur, not exposing it on the API.
Only when a request is used as Future<T> this exception is exposed as required by the interface.

eatonphil commented on src/java/src/com/tigerbeetle/AccountsBatch.java:
```diff
@@ -28,36 +27,40 @@ public AccountsBatch(Account[] accounts) {
         // Make sure the completion handler is giving us valid data
         if (bufferLen % Account.Struct.SIZE != 0)
-            throw new RequestException(RequestException.Status.INVALID_DATA_SIZE);
+            throw new AssertionError(
```
Why make up a new AssertionError instead of using actual assertions?

https://docs.oracle.com/javase/7/docs/technotes/guides/language/assert.html#:~:text=An%20assertion%20is%20a%20statement,than%20the%20speed%20of%20light.

batiati replied:
Two reasons:

- It evaluates Expression1 and if it is false throws an AssertionError with **no detail message.**
- To add more information to the stack trace with an exception in our package `com.tigerbeetle.AssertionError`

I think we should inherit this exception from `java.lang.AssertionError` to make more sense for someone who's expecting an AssertionError.

batiati replied:
Let me check one thing... I didn't know that assertions could be disabled on the JVM, and we don't want this on ours assertions. I'm not sure if inheritance is going to be a good idea here.

batiati replied:
I had to do a quick research on this matter, to be honest, I knew very little about the `assert` keyword. 

To summarize it:

~It evaluates Expression1 and if it is false throws an AssertionError with no detail message.~
I was wrong about that, there is a syntax-sugar `assert x > 0 : "Error message!"`.

But anyway, it seems that `assert` has an undesirable semantics:

From [this post](https://www.yegor256.com/2016/06/17/dont-use-java-assertions.html):
"This is the first difference between exceptions and assertions. Exceptions will always be thrown, while assertions are not enabled by default. They are supposed to be turned on during testing and turned off in production. "

Inheritance would work fine here, we can do `throw new AssertionException()` even with assertions disabled, it is only the `assert` keyword that is turned off, not the AssertionException itself.

batiati committed:
AssertionError now inherits from java.lang.AssertionError

Rationale:
- We can filter logs and track this kind of error by our classname com.tigerbeetle.AssertionError
- Log frameworks and analysis tools can handle this error like any other generic java.lang.AssertionError
